### PR TITLE
Use future rather than promise MODCR-94

### PR DESCRIPTION
**Pull request by adamdickmeiss**

### CODE REVIEW

File: src/main/java/org/folio/coursereserves/util/CRUtil.java [hunk 1]
```diff
@@ -116,86 +116,50 @@ public static Map<String, String> getTextAcceptHeaders() {
...
-      }
+      return Future.failedFuture("Must provide item id or item barcode to populate copied items");
+    }
+    return itemIdFuture.compose(itemIdRes -> {
+      reserve.setItemId(itemIdRes);
+      String retrievedItemId = itemIdRes;
+      logger.info("Looking up information for item {} from inventory module",
+          retrievedItemId);
+      return lookupItemHoldingsInstanceByItemId(retrievedItemId, okapiHeaders, context)
+          .map(inventoryRes -> {
```
Comment (julianladisch) [inline src/main/java/org/folio/coursereserves/util/CRUtil.java:145]: This additional nesting can be avoided:

```
      return lookupItemHoldingsInstanceByItemId(retrievedItemId, okapiHeaders, context);
    }).map(inventoryRes -> {
```

File: src/main/java/org/folio/coursereserves/util/CRUtil.java [hunk 2]
```diff
@@ -116,86 +116,50 @@ public static Map<String, String> getTextAcceptHeaders() {
...
-              promise.complete(inventoryRes.result());
-            } catch(Exception e) {
-              promise.fail(e);
-            }
-          }
-        });
-      }
+      return Future.failedFuture("Must provide item id or item barcode to populate copied items");
+    }
+    return itemIdFuture.compose(itemIdRes -> {
```
Comment (julianladisch) [inline src/main/java/org/folio/coursereserves/util/CRUtil.java:139]: Moving all the code for the itemIdFuture into a
`private static Future<String> getItemId(Reserve reserve, Map<String, String> okapiHeaders, Context context)`
method might be more readable.

File: src/main/java/org/folio/coursereserves/util/CRUtil.java [hunk 3]
```diff
@@ -299,211 +263,141 @@ public static void populateReserveCopiedItemFromJson(Reserve reserve, JsonObject
...
-              } else {
-                JsonObject instanceJson = instanceRes.result();
-                result.put("instance", instanceJson);
-                logger.info("Inventory lookup complete");
-                promise.complete(result);
-              }
-            });
-          }
+    return makeOkapiRequest(context.owner(), okapiHeaders, ITEMS_ENDPOINT + "/" + itemId,
+        HttpMethod.GET, null, null, 200).compose(itemJson -> {
```
Comment (julianladisch) [inline src/main/java/org/folio/coursereserves/util/CRUtil.java:268]: For me it's more readable if there's a line break before each `.compose` and `.map`.

File: src/main/java/org/folio/coursereserves/util/CRUtil.java [hunk 4]
```diff
@@ -299,211 +263,141 @@ public static void populateReserveCopiedItemFromJson(Reserve reserve, JsonObject
...
+      JsonObject result = new JsonObject();
+      String holdingsId = itemJson.getString("holdingsRecordId");
+      result.put("item", itemJson);
+      logger.info("Making request for holdings at {}/{}", HOLDINGS_ENDPOINT, holdingsId);
+      return makeOkapiRequest(context.owner(), okapiHeaders, HOLDINGS_ENDPOINT + "/" + holdingsId,
+          HttpMethod.GET, null, null, 200).compose(holdingsJson -> {
```
Comment (julianladisch) [inline src/main/java/org/folio/coursereserves/util/CRUtil.java:274]: Additional nesting can be avoided:
```
      return makeOkapiRequest(context.owner(), okapiHeaders, HOLDINGS_ENDPOINT + "/" + holdingsId,
          HttpMethod.GET, null, null, 200);
    }).compose(holdingsJson -> {
```

File: src/main/java/org/folio/coursereserves/util/CRUtil.java [hunk 5]
```diff
@@ -299,211 +263,141 @@ public static void populateReserveCopiedItemFromJson(Reserve reserve, JsonObject
...
-            } catch(Exception e) {
-              promise.fail(e);
-            }
-          });
-        }
-      } catch(Exception e) {
-        promise.fail(e);
-      }
+    return makeOkapiRequest(context.owner(), okapiHeaders, userPath, HttpMethod.GET,
+        null, null, 200).compose(userRes -> {
```
Comment (julianladisch) [inline src/main/java/org/folio/coursereserves/util/CRUtil.java:291]: For me it's more readable if there's a line break before each `.compose` and `.map`.

File: src/main/java/org/folio/coursereserves/util/CRUtil.java (hunk 2 above)
Comment (adamdickmeiss) [reply src/main/java/org/folio/coursereserves/util/CRUtil.java:139]: Agreed.

File: src/main/java/org/folio/coursereserves/util/CRUtil.java (hunk 4 above)
Comment (adamdickmeiss) [reply src/main/java/org/folio/coursereserves/util/CRUtil.java:274]: Yep. By defining result earlier.

File: src/main/java/org/folio/coursereserves/util/CRUtil.java [hunk 6]
```diff
@@ -299,211 +263,141 @@ public static void populateReserveCopiedItemFromJson(Reserve reserve, JsonObject
...
+      JsonObject result = new JsonObject();
+      result.put("user", userRes);
+      String groupId = userRes.getString("patronGroup");
+      String groupPath = "/groups/" + groupId;
+      return makeOkapiRequest(context.owner(), okapiHeaders, groupPath, HttpMethod.GET,
+          null, null, 200).map(groupRes -> {
```
[the remark this answers is not in the document]
Comment (adamdickmeiss) [reply src/main/java/org/folio/coursereserves/util/CRUtil.java:297]: yep

**Comment by adamdickmeiss** [comment]
Thx for looking at this PR, @julianladisch

File: src/main/java/org/folio/coursereserves/util/CRUtil.java [hunk 7]
```diff
@@ -515,202 +416,153 @@ public static Future<JsonObject> lookupItemByBarcode(String barcode,
...
+      }
+      String termId = courselisting.getTermId();
+      String courseTypeId = courselisting.getCourseTypeId();
+      String locationId = courselisting.getLocationId();
+      String servicepointId = courselisting.getServicepointId();
+      Future<Term> termFuture;
+      Future<CourseType> coursetypeFuture;
+      Future<JsonObject> locationFuture;
+      Future<JsonObject> servicePointFuture;
+      if (expandTerm && termId != null) {
```
Comment (julianladisch) [inline src/main/java/org/folio/coursereserves/util/CRUtil.java:491]: All these multiple `if (expandTerm` can be avoided by using
```
    if (! expandTerm) {
      return Future.succeededFuture(courselisting);
    }
```
once.

Comment (adamdickmeiss) [reply src/main/java/org/folio/coursereserves/util/CRUtil.java:491]: OK. But we can do even better. expandTerm is always true. I'll remove the parameter and reduce complexity that way.

File: src/main/java/org/folio/coursereserves/util/CRUtil.java [hunk 8]
```diff
@@ -515,202 +416,153 @@ public static Future<JsonObject> lookupItemByBarcode(String barcode,
...
-                courselisting.setLocationObject(locationObjectFromJson(locationFuture.result()));
-              }
-              if(servicePointFuture.succeeded()) {
-                courselisting.setServicepointObject(servicepointObjectFromJson(servicePointFuture.result()));
-              }
-              promise.complete(courselisting);
-            } catch(Exception e) {
-              promise.fail(e);
+        termFuture = Future.failedFuture("No lookup");
+      }
```
Comment (julianladisch) [inline src/main/java/org/folio/coursereserves/util/CRUtil.java:495]: Moving related code to one place:
```
      List<Future> futureList = new ArrayList<>();
      if (termId != null) {
        futureList.add(
            lookupTerm(termId, okapiHeaders, context)
            .map(CRUtil::termObjectFromTerm)
            .onSuccess(courselisting::setTermObject)
        );
      }
      if (courseTypeId != null) {
        futureList.add(
          lookupCourseType(courseTypeId, okapiHeaders, context)
          .map(CRUtil::courseTypeObjectFromCourseType)
          .onSuccess(courselisting::setCourseTypeObject)
      )
      ...
      return CompositeFuture.join(futureList)
          .recover(x -> Future.succeededFuture())  // is this still needed?
          .map(courselisting);
```

Comment (adamdickmeiss) [reply src/main/java/org/folio/coursereserves/util/CRUtil.java:495]: Good idea. Will do (probably not today).

**Review by julianladisch — APPROVED** [review]
Good code cleanup, thanks!

File: src/main/java/org/folio/coursereserves/util/Util.java [hunk 9]
```diff
@@ -8,4 +8,18 @@ public static String logAndSaveError(Throwable err, Logger logger) {
...
+   * @param limitQuery simple field=value query
+   * @return combined CQL query.
+   */
+  public static String getQueryWithLimit(String userQuery, String limitQuery) {
+    if (userQuery == null || userQuery.isEmpty()) {
+      return limitQuery;
+    } else {
+      return limitQuery + " AND (" + userQuery + ")";
+    }
+  }
```
Comment (julianladisch) [inline src/main/java/org/folio/coursereserves/util/Util.java:24]: The APIs have a limit=10 parameter, this makes the limit in `limitQuery` somewhat misleading for me.
If both sub-queries are put into braces people can use OR in both of them.
How about a general-purpose method with each sub-query being optional?

```
  public static String cqlAnd(String subQuery1, String subQuery2) {
    if (StringUtils.isBlank(subQuery1)) {
      if (StringUtils.isBlank(subQuery2)) {
        return "cql.allRecords=1";
      } else {
        return subQuery2;
      }
    } else {
      if (StringUtils.isBlank(subQuery2)) {
        return subQuery1;
      } else {
        return "(" + subQuery1 + ") AND (" + subQuery2 + ")";
      }
    }
  }
```

Comment (adamdickmeiss) [reply src/main/java/org/folio/coursereserves/util/Util.java:24]: People can not put the limitQuery in braces or anything like that. It's a simple field=<value> where the value is a path component , like listening_id. Therefore the implementation I chose was the minimal to support that.

**Review by julianladisch — APPROVED** [review]
Looks good, thanks!

**Pull request merged**